Feature ebtree views

davisp opened the pull request:
## Overview

Full map/reduce implementation using ebtree's

## Testing recommendations

`make check`

## Related Issues or Pull Requests

These are a few other attempts at reduce functions with various tradeoffs and limitations.

#2456
#2984 
#3018 

## Checklist

- [x] Code is written and works correctly
- [x] Changes are covered by tests
- [x] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

nickva commented on src/couch_views/src/couch_views_indexer.erl:
```diff
@@ -589,6 +613,51 @@ fail_job(Job, Data, Error, Reason) ->
+                Str = io_lib:format("~s:~w", [Id, Value]),
+                [Str | Acc]
+            end, [], KVs),
+            Msg = "XKCD VIEW STATS: " ++ string:join(lists:reverse(Strs),  " "),
```
Debug log

davisp replied:
Whoops. That whole commit is debugging. Will remove it.

davisp replied:
And its like it never even happened.

iilyak commented on src/couch_views/test/couch_views_indexer_test.erl:
```diff
@@ -388,8 +384,8 @@ multiple_design_docs(Db) ->
     meck:new(couch_views_fdb, [passthrough]),
-    meck:expect(couch_views_fdb, write_doc, fun(TxDb, Sig, ViewIds, Doc) ->
-        meck:passthrough([TxDb, Sig, ViewIds, Doc])
+    meck:expect(couch_views_fdb, update_views, fun(TxDb, Mrst, Docs) ->
```
This is confusing.  We create an `expect` for an `update_views`, but use `meck:num_calls/3` on `write_doc`.

davisp replied:
Good catch. write_doc was renamed and I forgot to update the test. Will update with a fix.

davisp replied:
Fixed.

garrensmith reviewed (changes requested):
Generally looking good. The big issue I have is the tree related functions in `couch_views_fdb`, I think they should be in a new module. I'm not sure why we need the `id_tree` though, it wasn't obvious to me what that buys us other than everything is a btree?

Finally, It would be nice to see some performance results and how it compares to the old map implementation.

davisp commented on rel/overlay/etc/default.ini:
```diff
@@ -337,6 +337,12 @@ iterations = 10 ; iterations for password hashing
 ; The maximum allowed value size emitted from a view for a document (in bytes)
 ;value_size_limit = 64000
 ;
+; The order of B+Tree nodes used by the id btree
```
I've just changed the comment to say "maximum size" rather than using the term from literature.

davisp commented on src/couch_views/test/couch_views_red_test.erl:
```diff
@@ -0,0 +1,764 @@
+-define(TDEFI(A), {atom_to_list(A), fun A/0}).
+
+
+with(Tests) ->
```
Good catch. I didn't realize till after that I could include a `.hrl` from a test directory. Have updated to remove these macros.

davisp replied:
Updated to use fabric2_test.hrl.

nickva commented on rel/overlay/etc/default.ini:
```diff
@@ -337,6 +337,12 @@ iterations = 10 ; iterations for password hashing
 ; The maximum allowed value size emitted from a view for a document (in bytes)
 ;value_size_limit = 64000
 ;
+; The order of B+Tree nodes used by the id btree
+;id_btree_node_size = 100
+;
+; The order of B+Tree nodes used by view btrees
+;view_btree_node_size = 100
```
What are some trade-offs for view be when setting this higher or lower. Say what would setting to 10 do vs say 1000

davisp replied:
There's a bunch of different variables that we'll want to balance:

* Size of resultant FDB values
* Height of the tree affects performance (lower limits cause taller trees)
* Number of FDB KVs touched in a transaction
* Speed of reductions
* Speed of iterating over maps
* Lots of other things

100 is basically a hand wave guess that seems to work well in practice. But I didn't want it hard coded so exposed the knobs.

nickva commented on src/couch_views/src/couch_views_indexer.erl:
```diff
@@ -110,6 +112,10 @@ init() ->
             fail_job(Job, Data, db_deleted, "Database was deleted");
         Error:Reason  ->
+            Stack = erlang:get_stacktrace(),
+            Fmt = "Error building view for ddoc ~s in ~s: ~p:~p ~p",
```
Good idea to add this

nickva commented on src/couch_views/src/couch_views_indexer.erl:
```diff
@@ -102,6 +102,8 @@ init() ->
         update_stats => #{}
     },
 
+    process_flag(sensitive, false),
```
Why is this needed?

davisp replied:
Ah derps. That was leftover from testing when trying to trace when aegis was accidentally setting it to true. The other spot has been fixed so this can be removed now.

davisp replied:
Removed.

davisp commented on src/couch_views/src/couch_views_fdb.erl:
```diff
@@ -244,200 +290,284 @@ clear_index(Db, Signature) ->
 
 
-update_row_count(TxDb, Sig, ViewId, Increment) ->
+open_view_tree(TxDb, Sig, _Lang, View) ->
```
I've gone and created `couch_views_trees.erl` which contains all of the tree reading and writing related functions.

davisp commented on src/couch_views/src/couch_views_fdb.erl:
```diff
@@ -244,200 +290,284 @@ clear_index(Db, Signature) ->
-row_count_key(DbPrefix, Sig, ViewId) ->
-    Key = {?DB_VIEWS, ?VIEW_INFO, ?VIEW_ROW_COUNT, Sig, ViewId},
-    erlfdb_tuple:pack(Key, DbPrefix).
+    lists:foldl(fun(Doc, InfoAcc2) ->
```
This was changed so the anonymous function is named and includes this function head match.

davisp commented on src/couch_views/src/couch_views_fdb.erl:
```diff
@@ -126,92 +128,136 @@ set_update_seq(TxDb, Sig, Seq) ->
+    } = gather_update_info(Tx, Mrst, Docs),
+
+    % Generate a list of Keys to delete and Rows to insert from a map
+    UpdateBTree = fun(BTree, Map) ->
```
Done.

davisp commented on src/couch_views/test/couch_views_red_test.erl:
```diff
@@ -0,0 +1,764 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
```
Added your tests.

davisp commented on src/couch_views/src/couch_views_fdb.erl:
```diff
@@ -126,92 +128,136 @@ set_update_seq(TxDb, Sig, Seq) ->
+    UpdateBTree = fun(BTree, Map) ->
+        {ToRemove, ToInsert} = maps:fold(fun(Key, Value, {Keys, Rows}) ->
+            case Value of
+                DeleteRef -> {[Key | Keys], Rows};
```
I don't like that. The loop splits into two groups whcih are then acted on separately. Acting on some keys while reserving others seems more awkward than a split.

davisp commented:
@garrensmith I've updated things to use a `couch_views_trees` module for all the tree related functions. On using ebtree for the `id_tree` it simplifies things greatly as we don't have to maintain a separate set of encoding/decoding functions.

I'll run through a 1M doc update again locally and post a link to the scripts and results. Last I checked it was roughly 2-5 times faster than the existing view implementation but I'll double check I've not messed anything up.

nickva reviewed (approved):
I built views with some built-in reducers (`_count`, `_approx_count_distinct`) then a few custom ones. Tested view updates, queries, group_levels, descending=true|false, start_key/end_key. Set `id_btree_node_size` and `view_btree_node_size` to smaller values like 6, 10 and 20 and it worked.

Good idea to enable some more elixir tests and bring in @garrensmith's work

There are also a few basic tests that use reduce that were commented out, we can uncomment them now as well:

https://github.com/apache/couchdb/blob/fa3c3c8dd3d5b77371cd74dfbe149623797160ad/test/elixir/test/basics_test.exs#L195

https://github.com/apache/couchdb/blob/fa3c3c8dd3d5b77371cd74dfbe149623797160ad/test/elixir/test/basics_test.exs#L208

One thing that changed from 3.x is how we handle errors:

```
{
    "rows": [
        {
            "error": "reduce_overflow_error",
            "key": null,
            "reason": "Reduce output must shrink more rapidly: input size: 671 output size: 10003"
        }
    ]
}
```

vs

```
{
    "rows": [
        {
            "key": null,
            "value": {
                "reduce_overflow_error": "Reduce output must shrink more rapidly: input size: 779 output size: 10003"
            }
        }
    ]
}
```

However this can be done in a separate PR.

Nice work!

+1

davisp commented:
For the reduce error test I'm working on a second PR that re-enables most of the Elixir test suite that now passes. This error is flagged there so we can keep track of it that way.

nickva commented:
@davisp that sounds good!

davisp commented:
For reference, that PR is #3175 but has this PR branch as its base.

davisp commented:
These are three replicas building a simple view against 1M randomized documents. Each test is local to my laptop so take that for what its worth.

main: 7m40s, 8m40s, 9m8s
feature-ebtree-views: 8m46s, 9m36s, 8m47s

I'm fairly surprised to see the main branch looking a bit faster with that first run at 7m40s. When we've run similar tests against FoundationDB clusters that version is significantly slower than the ebtree branch. My best guess at this point is that its due to touching a significantly larger number of fdb key/value pairs since each row is separate.

davisp commented:
Also, the scripts for generating those numbers are here:

https://gist.github.com/davisp/3863ab02fb5def733f1a6991651840f9

garrensmith commented:
I did a quick group_level benchmarking this morning to compare it to the map + ebtree work I built-in #3018. The benchmark builds a view with keys that are a group_level = 3. Group_levels 1 and 2 will group results. 

The full ebtree builds a lot faster, which is great. And queries for group_level = 1 and group_level = 2 are very similar. But when doing a reduce query for group_level = 3 the full ebtree is noticeably slower.

Example results for 300K docs:
full ebtree (this PR):
```
Insert Time  71435.896557 ms
Querying:
Group 1 Time 12.577319 ms
Result Length 5
Group 2 Time 20.666643 ms
Result Length 20
Group 3 Time 3712.46744 ms
Result Length 300002
```

Map + ebtree reduce (PR #3018):
```
Insert Time  362359.221368 ms
Querying:
Group 1 Time 13.821064 ms
Result Length 5
Group 2 Time 23.708709 ms
Result Length 20
Group 3 Time 2672.549299 ms
Result Length 300002
```

Code for the benchmarking is here https://gist.github.com/garrensmith/9bd090cb00f6b641e58a3c490b2300e7
Same warning as Paul's this is run locally on my machine.

davisp commented:
@garrensmith That seems reasonably close. I'd suspect that the difference there is likely the number of times that ebtree-only has to call out to the reduce function to process individual groups. Though 80k rows/sec vs 110k rows/sec is honestly a bit surprising given that we're running the reduce function 300k times roughly.

Are you concerned about the difference there? It seems not a deal breaker given that the full ebtree PR supports 100% API compatibility for reduce functions rather than just pre-specified group levels.

garrensmith commented:
Hi Paul,

That isnt correct. The benchmark is with my branch that uses ebtree for reduce along with the old map index. So it also supports the full reduce api. I found it surprising that it was faster than this PR specially since it uses a much older version of ebtree. I think it would be worth investigating why this work is slower than my PR.

________________________________
From: Paul J. Davis <notifications@github.com>
Sent: Monday, September 28, 2020 6:37:36 PM
To: apache/couchdb <couchdb@noreply.github.com>
Cc: garren smith <garren.smith@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [apache/couchdb] Feature ebtree views (#3164)


@garrensmith<https://github.com/garrensmith> That seems reasonably close. I'd suspect that the difference there is likely the number of times that ebtree-only has to call out to the reduce function to process individual groups. Though 80k rows/sec vs 110k rows/sec is honestly a bit surprising given that we're running the reduce function 300k times roughly.

Are you concerned about the difference there? It seems not a deal breaker given that the full ebtree PR supports 100% API compatibility for reduce functions rather than just pre-specified group levels.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/couchdb/pull/3164#issuecomment-700147946>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AABL2AWHBLMDV55TTBQXRTDSIC3VBANCNFSM4RSQN3FQ>.

davisp commented:
Oh derps. I was thinking of the pre-computed group levels. Somewhat intriguing. I'll investigate tomorrow and post back.

davisp commented:
@garrensmith A bit of profiling showed that I was doing a bit of extra work compared to #3018 so there were a few simple optimizations to be made. Namely:

1. Avoid calculating unused reductions while reading from a reduce view (coincidentally an optimization that exists even in the legacy view engine
2. Avoid some extra work in couch_query_servers that I re-used instead of duplicating the functionality of this PR.

For 1, the #3018 PR achieves this same end by having a separate ebtree instance for each reduce function on the view. For 2, its similar other than the size calculation is done on at write time.

![Screen Shot 2020-09-29 at 3 18 02 PM](https://user-images.githubusercontent.com/19929/94612436-b7e49f80-0268-11eb-97e1-549b8b5ec80b.png)

Updated numbers generated locally. I ported the Node.js script to Python [1] and removed deleting the database part. I then ran it four times. Once to generate the index, and then three times to gather numbers after everything was warmed up in cache to try and get an equal footing for each replica.

There's still about 200ms difference between the two implementations. However, based on profiling both implementations it would appear that a good chunk of that is due to #3018 performing significantly fewer invocations of collation which is a bug in a previous version of ebtree that has since been fixed [2]. I also went fixed the lack of JSON collation in #3018 [3] and that appears to have made it faster.

I'm fairly confident I've covered the differences given the similar algorithmic properties of the implementations. Let me know if you're OK merging this now.

[1] https://gist.github.com/davisp/48303dcd877b31632fb75055d6758129
[2] https://github.com/apache/couchdb/commit/f8fdf9721e2ac932022065bc075301641568d67c
[3] https://github.com/cloudant/couchdb/compare/fdb-btree-reduce...apache:fdb-btree-reduce-davisp

garrensmith commented:
Nice work. Thanks for checking that. I’m away until next week so I can’t give a final review until then. I can’t remember if anyone else has approved this or not. If they have then go ahead and merge.

________________________________
From: Paul J. Davis <notifications@github.com>
Sent: Tuesday, September 29, 2020 10:58:32 PM
To: apache/couchdb <couchdb@noreply.github.com>
Cc: garren smith <garren.smith@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [apache/couchdb] Feature ebtree views (#3164)


@garrensmith<https://github.com/garrensmith> A bit of profiling showed that I was doing a bit of extra work compared to #3018<https://github.com/apache/couchdb/pull/3018> so there were a few simple optimizations to be made. Namely:

  1.  Avoid calculating unused reductions while reading from a reduce view (coincidentally an optimization that exists even in the legacy view engine
  2.  Avoid some extra work in couch_query_servers that I re-used instead of duplicating the functionality of this PR.

For 1, the #3018<https://github.com/apache/couchdb/pull/3018> PR achieves this same end by having a separate ebtree instance for each reduce function on the view. For 2, its similar other than the size calculation is done on at write time.

[Screen Shot 2020-09-29 at 3 18 02 PM]<https://user-images.githubusercontent.com/19929/94612436-b7e49f80-0268-11eb-97e1-549b8b5ec80b.png>

Updated numbers generated locally. I ported the Node.js script to Python [1] and removed deleting the database part. I then ran it four times. Once to generate the index, and then three times to gather numbers after everything was warmed up in cache to try and get an equal footing for each replica.

There's still about 200ms difference between the two implementations. However, based on profiling both implementations it would appear that a good chunk of that is due to #3018<https://github.com/apache/couchdb/pull/3018> performing significantly fewer invocations of collation which is a bug in a previous version of ebtree that has since been fixed [2]. I also went fixed the lack of JSON collation in #3018<https://github.com/apache/couchdb/pull/3018> [3] and that appears to have made it faster.

I'm fairly confident I've covered the differences given the similar algorithmic properties of the implementations. Let me know if you're OK merging this now.

[1] https://gist.github.com/davisp/48303dcd877b31632fb75055d6758129
[2] f8fdf97<https://github.com/apache/couchdb/commit/f8fdf9721e2ac932022065bc075301641568d67c>
[3] cloudant/couchdb@fdb-btree-reduce...apache:fdb-btree-reduce-davisp<https://github.com/cloudant/couchdb/compare/fdb-btree-reduce...apache:fdb-btree-reduce-davisp>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/couchdb/pull/3164#issuecomment-700982451>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AABL2AREMBWCDZ7B4AFEHSLSIJC7RANCNFSM4RSQN3FQ>.

rnewson reviewed (approved):
Thanks Garren.

I've read all the above, and the code (both current and previous versions as its evolved), I'm happy to move forward.
